### PR TITLE
Infrastructure for testing inter project imports and exports

### PR DESCRIPTION
**Pull request by JaroslavTulach**
### Pull Request Description

Infrastructure for testing **inter** project imports with few succeeding tests.

### Checklist

Please ensure that the following checklist has been satisfied before submitting the PR:

- [x] All code follows the
      [Scala](https://github.com/enso-org/enso/blob/develop/docs/style-guide/scala.md),
- All code has been tested:
  - [x] Unit tests have been written where possible.

### CODE REVIEW

**Comment by JaroslavTulach** [comment]
Trying to test how **inter** project imports work I realized I don't know what approch to copy!? There is:
- [ImportsTest.scala](https://github.com/enso-org/enso/blob/326813900585e0f54d3d4c9e1eede4591c31e30b/engine/runtime/src/test/scala/org/enso/interpreter/test/semantic/ImportsTest.scala#L5) by @hubertp 
- [ImportExportTest.scala](https://github.com/enso-org/enso/blob/326813900585e0f54d3d4c9e1eede4591c31e30b/engine/runtime/src/test/scala/org/enso/compiler/test/semantic/ImportExportTest.scala#L22) by @Akirathan 

None of them (as far as I can tell) shows how to test **inter** project imports. All the imports and exports are happening inside of a single project. This PR is my attempt to show how a testing infrastructure for **inter** project exports and imports could look like.

Of course, as Pavel told me last week, the problem is `EditionManager` - it is quite hardcoded and not easy to be extended. To overcome that I decided to expand `micro-distribution` with `Test` namespace modules. As the `ImportsTest` is using `micro-distribution` (good, as debugging imports on real `Standard.Base` is too huge task to swallow), one can import these `Test` modules in.

Turns out that _the most important trick_ is to put the `Test` libraries into `micro-distribution`! Then one can do any kind of testing with it. 51f6cde shows how to use this from inside of [ImportExportTest](https://github.com/enso-org/enso/pull/6840/commits/51f6cde1cc82a8b51d296bd662e889245fc89677).

Is this OKeyish testing infrastructure Hubert, Pavel?

**Comment by JaroslavTulach** [comment]
OK, 18b5c29 probably [wasn't good idea](https://github.com/enso-org/enso/actions/runs/5078139380/jobs/9122286622?pr=6840#step:10:4115). There are two failing tests:
- `org.enso.interpreter.test.semantic.ImportsTest`
- `org.enso.interpreter.test.semantic.PolyglotTest`

I need to change them to get just the original `getMessage` and not the _human readable_ one. Done in [9fbba8b](https://github.com/enso-org/enso/pull/6840/commits/9fbba8be10e7c620ca529472e84971f3a9efcced)

File: engine/runtime/src/test/scala/org/enso/interpreter/test/InterpreterException.scala [hunk 1]
```diff
@@ -3,23 +3,26 @@ package org.enso.interpreter.test
...
 
 case class InterpreterException(
-  @transient polyglotException: PolyglotException
+  @transient polyglotException: PolyglotException,
+  @transient output: Option[Any]
 ) extends Throwable
     with Serializable {
-  override def getMessage: String = polyglotException.getMessage
+  override def getMessage: String =
+    polyglotException.getMessage + output.map("\n" + _.toString()).getOrElse("")
```
Comment (Akirathan) [inline engine/runtime/src/test/scala/org/enso/interpreter/test/InterpreterException.scala]: This is very cool, I missed this feature a lot.

**Review by Akirathan — APPROVED** [review]
LGTM. I can't wait to add more libraries and more tests. This will make the transition to the better import system much smoother.

**Comment by radeusgd** [comment]
> the problem is `EditionManager` - it is quite hardcoded and not easy to be extended.

The [edition configuration](https://enso.org/docs/developer/enso/libraries/editions.html#editions) can be extended by defining a custom edition for the test project.

I'm not sure if there are good examples how to do it in a test project and I imagine there could be some obstacles to working with editions in a test environment (although ideally there shouldn't be any obstacles), but it should be very well doable to define custom editions for testing purposes.

**Review by hubertp — APPROVED** [review]

File: test/micro-distribution/lib/Test/Logical_Export/0.0.0-dev/src/Impl/Internal.enso [hunk 2]
```diff
@@ -0,0 +1,4 @@
+type Internal
+    Impl
+
+    describe self = "Internal"
```
Comment (hubertp) [inline test/micro-distribution/lib/Test/Logical_Export/0.0.0-dev/src/Impl/Internal.enso]:
```suggestion
    describe self = "Internal"

```

**Review by GregoryTravis — APPROVED** [review]

**Comment by JaroslavTulach** [comment]
>  but it should be very well doable to define custom editions for testing purposes.

I don't think there is any `EnsoLanguage` "option" to **mock in** custom logic. There is the `LANGUAGE_HOME_OVERRIDE` and `EDITION_OVERRIDE` but those are already used by the test. 

Thanks for the [edition link](https://enso.org/docs/developer/enso/libraries/editions.html#editions):
- _"curated Edition file will guarantee that all libraries are compatible which simplifies version resolution"_ - yup that's a solution to the incompatibilities problem - just like Linux distribution like Ubuntu 22.0 guarantees all packages on official update centers are compatible
-  "if current edition defines a library ... that was already defined in the parent edition, ... the current edition takes precedence." - relates to [Know it better! "paradox"](http://wiki.apidesign.org/wiki/SuperVsInner)  - but correct decision in this case - curators of editions shall not be [clueless](http://wiki.apidesign.org/wiki/Cluelessness)
- "you can shadow a repository name" - why? It can only create confusion. It should rather be forbidden to override repository definition when it has no effect on parent edition behavior
- "custom edition search paths are ... defined by the `ENSO_EDITION_PATH`" - e.g. one can have special edition for testing, but it cannot be _"virtual"_, it has to be on disk somewhere - similar to using `micro-distribution` as proposed by this PR
-  if the project has `prefer-local-libraries` set to `true` and if any directory on the `ENSO_LIBRARY_PATH` contains `Foo/Bar` - I guess we could set the `ENSO_LIBRARY_PATH` env variable in the test or create an option for `EnsoLanguage` to allow an engine to be configured as such. Then the `Test` libraries could be next to the other libraries used by the `ImportsTest.scala`

It is always valuable to be able to see the original design ideas.

**Comment by JaroslavTulach** [comment]
Reviewers, please vote. _"Thumbs up"_ it is OK as it is. _"Thumbs down"_ to investigate the following change:

>  if the project has `prefer-local-libraries` set to `true` and if any directory on the `ENSO_LIBRARY_PATH` contains `Foo/Bar`

Should I move the `Test/Logical_Export` project away from  `micro-distribution` and find a way to set `ENSO_LIBRARY_PATH` to point to [engine/runtime/src/test/resources/](https://github.com/enso-org/enso/pull/6840/files#diff-e8c34400058474ca9ad1bc60cb23ffe1014105724d03c859ad6f626ed47e49c7) so all the sibling projects can be loaded by the `ImportsTest.scala` & other properly configured tests?

File: engine/runtime/src/test/scala/org/enso/interpreter/test/InterpreterException.scala (hunk 1 above)
Comment (JaroslavTulach) [reply engine/runtime/src/test/scala/org/enso/interpreter/test/InterpreterException.scala]: Needed few more fixes: 9fbba8b, but yes, I believe it is going to be useful.

**Comment by Akirathan** [comment]
> Should I move the Test/Logical_Export project away from micro-distribution and find a way to set ENSO_LIBRARY_PATH to point to [engine/runtime/src/test/resources/](https://github.com/enso-org/enso/pull/6840/files#diff-e8c34400058474ca9ad1bc60cb23ffe1014105724d03c859ad6f626ed47e49c7) so all the sibling projects can be loaded by the ImportsTest.scala & other properly configured tests?

I think that the current structure is sufficient. It allows us to (easily) create other libraries just by adding them to `test/micro-distribution/lib/Test`. There is no need to try to hack `EditionManager` just yet.

**Review by jdunkerley — APPROVED** [review]

**Pull request merged**